app/models/custom_button_set.rb

```diff
@@ -28,6 +28,14 @@ def update_children
     replace_children(children)
   end
 
+  def self.update_group_index(ids)
+    ids.each.with_index(1) do |id, i|
+      button_set = CustomButtonSet.find(id)
+      button_set.set_data[:group_index] = i
+      button_set.save!
+    end
```

Comment:
Is it possible to look up all the `ButtonSet` records in a single query instead of `N+1` style?
Not sure if we need to ensure that all of the ids are valid for `CustomButtonSet`. (punting for now)


```suggestion
    CustomButtonSet.where(:id => ids).each do |button_set|
      button_set.set_data[:group_index] = ids.index(button_set.id) + 1
      button_set.save!
    end
```

Reply:
Agree with Keenan here that we should eliminate the N+1.  I think it's trivial to add a check that the requested ids all exist and are uniq, though, so I think we should do that.  Additionally, this code indexes the ids up front, to keep the clearer logic.

```suggestion
    sets = CustomButtonSet.where(:id => ids).index_by(&:id)
    raise ArgumentError, "not all requested ids found" unless sets.keys.sort == ids.sort
        
    ids.each_with_index do |id, i|
      button_set = sets[id]
      button_set.set_data[:group_index] = i
      button_set.save!
    end
```

Reply:
Agree with Keenan here that we should eliminate the N+1.  I think it's trivial to add a check that the requested ids all exist and are uniq, though, so I think we should do that.  Additionally, this code indexes the ids up front, to keep the clearer logic.

```suggestion
    sets = CustomButtonSet.where(:id => ids).index_by(&:id)
    raise ArgumentError, "not all requested ids found" unless sets.keys.sort == ids.sort
        
    ids.each_with_index do |id, i|
      button_set = sets[id]
      button_set.set_data[:group_index] = i
      button_set.save!
    end
```